ferretdb/ferretdb.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ferretdb
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
+	"github.com/FerretDB/FerretDB/internal/clientconn"
+	"github.com/FerretDB/FerretDB/internal/handlers/registry"
+	"github.com/FerretDB/FerretDB/internal/util/logging"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+	"github.com/FerretDB/FerretDB/internal/util/version"
+)
+
+// Config contains a PostgreSQL backend connection string.
+type Config struct {
+	PostgresURL string
+	TigrisURL   string
+}
+
+// FerretDB proxy.
+type FerretDB struct {
+	config   Config
+	mongoURL string
+}
+
+// New returns a new FerretDB.
+func New(conf Config) FerretDB {
+	return FerretDB{
+		config:   conf,
+		mongoURL: "mongodb://127.0.0.1:27017",
+	}
+}
+
+// GetConnectionString returns the backend connection string.
+func (fdb *FerretDB) GetConnectionString() string {
+	return fdb.mongoURL
+}
+
+// Run runs the FerretDB proxy as a library with:
+// * error level logging
+// * monitoring disabled
+// * handler PostgreSQL.
+func (fdb *FerretDB) Run(ctx context.Context, conf Config) error {
+	listenAddr := "127.0.0.1:27017"
+	proxyAddr := "127.0.0.1:37017"
+	mode := clientconn.NormalMode
+	handler := "pg"
+	testConnTimeout := time.Duration(0)
```

Comment:
`listenAddr`, `proxyAddr`, `mode` and `testConnTimeout` used only once in `NewListenerOpts` initialization. Why we need all those variables at all?

Reply:
> `listenAddr`, `proxyAddr`, `mode` and `testConnTimeout` used only once in `NewListenerOpts` initialization. Why we need all those variables at all?

To connect to the all the stuff the listener requires.

Reply:
Ok, I'll rephrase. Why do we need to introduce new variables that we don't use?

Reply:
they are used in the line 86:
```go
l := clientconn.NewListener(&clientconn.NewListenerOpts{
		ListenAddr:      listenAddr,
		ProxyAddr:       proxyAddr,
		Mode:            clientconn.Mode(mode),
		Handler:         h,
		Logger:          logger,
		TestConnTimeout: testConnTimeout,
	})
```

And it's better to have them on the top as a variables list to see them as settings.

Reply:
You can't treat those variables as settings, because you are required to change the code to set a new value. Settings are something you can pass from somewhere else.

Reply:
> You can't treat those variables as settings, because you are required to change the code to set a new value. Settings are something you can pass from somewhere else.

IMHO, as these variables are constant values right now, they should be placed on the top. 
I don't have that rule in mind that settings have to be passed from somewhere else. 
Settings are something that sets software up. And that's it.

Reply:
Usually, we place constants somewhere in the `const` section:
```go
const (
// ... 
)
```
If it's for future improvements, then we should move those variables to the `const` section.

Reply:
🆗